src/python/pants/backend/terraform/fmt.py

```diff
@@ -0,0 +1,68 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from dataclasses import dataclass
+from typing import Iterable, List, Type
+
+from pants.backend.terraform.target_types import TerraformSources
+from pants.core.goals.fmt import EnrichedFmtResult, LanguageFmtResults, LanguageFmtTargets
+from pants.core.goals.style_request import StyleRequest
+from pants.core.util_rules.source_files import SourceFiles, SourceFilesRequest
+from pants.engine.fs import Digest, Snapshot
+from pants.engine.internals.selectors import Get
+from pants.engine.rules import collect_rules, rule
+from pants.engine.unions import UnionMembership, UnionRule, union
+
+
+@dataclass(frozen=True)
+class TerraformFmtTargets(LanguageFmtTargets):
+    required_fields = (TerraformSources,)
+
+
+@union
+class TerraformFmtRequest(StyleRequest):
+    pass
+
+
+@rule
+async def format_terraform_targets(
+    terraform_fmt_targets: TerraformFmtTargets, union_membership: UnionMembership
+) -> LanguageFmtResults:
+    original_sources = await Get(
+        SourceFiles,
+        SourceFilesRequest(target[TerraformSources] for target in terraform_fmt_targets.targets),
+    )
+    prior_formatter_result = original_sources.snapshot
+
+    results: List[EnrichedFmtResult] = []
+    fmt_request_types: Iterable[Type[TerraformFmtRequest]] = union_membership.union_rules[
+        TerraformFmtRequest
+    ]
+    for fmt_request_type in fmt_request_types:
```

Comment:
Note: `terraform fmt` is currently the only expected formatter for Terraform files. However, I made it pluggable here in case users want to add their own custom formatters.

Reply:
Thanks. That's currently the sole way to add a formatted. FYI, this means that Terraform fmt will run concurrently with shfmt and Python formatters (but Python runs sequentially).